### ISSUE TITLE
lib.es6.d.ts is missing a method for CanvasRenderingContext2D interface 

### ISSUE THREAD

**Issue by nhrones** [Microsoft/TypeScript#19195]
lib.es6.d.ts is missing a 'fill' method for CanvasRenderingContext2D that takes a 'Path2D'

Please add the following to the CanvasRenderingContext2D interface:
`fill(path: Path2D, fillRule?: CanvasFillRule): void`

**Comment by mhegazy** [comment]
PRs welcomed. You can find more information about contributing lib.d.ts fixes at https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md#contributing-libdts-fixes.

**Comment by nhrones** [comment]
I don't understand this response? I'm NOT contributing! I'm only trying to
point out that your .d.ts files are broken, and causing many errors during
my compile.  None of the Typescript .d.ts files are complete with respect
to the canvasContext2D specification specifically relating to Path2D. There
are several omissions.

**Comment by RyanCavanaugh** [comment]
@nhrones we agree the definition files are incomplete. @mhegazy was outlining the process for contributing a fix for anyone who was interested in doing so.

**Comment by nhrones** [comment]
Is anyone taking any responsibility for the quality of this language?  The
Path2D issue in all .d.ts files has been known and ignored since December
2015 ...
https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwjV15_8gvjWAhUBZyYKHYU_Am4QFggnMAA&url=https%3A%2F%2Fstackoverflow.com%2Fquestions%2F34109921%2Fusing-path2d-in-a-typescript-project-is-not-resolved&usg=AOvVaw3AQecd8K3O5RuFJCetHmtq

I really don't understand the offer for me to fix .d.ts files?  Does this
mean anyone can modify the core  .d.ts files? If so, it is no wonder that
they are full of errors and omissions?

**Comment by RyanCavanaugh** [comment]
Why is software imperfect? Why do bugs exist? What has caused our resources to be woefully finite? Is imperfection our curse, or our penance? These questions exist, but until the hour of reckoning, the answers are shrouded from our view.

**Comment by nhrones** [comment]
So ... what you're telling me is 'it is best to avoid Typescript until the
team gets more resources!'
Message received .. you can close this issue.

**Comment by nhrones** [comment]
Okay. Looks like you came up with a fix in Typescript 2.6.
// @ts-ignore
Having compile errors due to incorrect/incomplete .d.ts?  Just add an ignore comment to your code.
Great fix. Now I know why your resources are so thin.

**Comment by RyanCavanaugh** [comment]
Reopening since this does need to get fixed - thanks for raising it. We'll fix it with the next round of DOM updates or hopefully someone can contribute a fix.

**Comment by nhrones** [comment]
Thanks, I keep updating my local declaration files, but each vs-code update replaces them. I used to know how to include an extension.d.ts, but the method to do this has changed so many times that I can't get any local files recognized. I don't use node_modules, and it looks like that is where code looks now?
If I had a tutorial on how to contribute, I would as I'm retired with lots of spare time.
By the way, there are 4 method declarations missing from interface CanvasRenderingContext2D that are related to Path2D:

```
clip(path: Path2D, fillRule?: CanvasFillRule): void;
fill(path: Path2D, fillRule?: CanvasFillRule): void;
isPointInPath(path: Path2D, x: number, y: number, fillRule?: CanvasFillRule): boolean;
isPointInStroke(path: Path2D, x: number, y: number): boolean;
```

**Comment by RyanCavanaugh** [comment]
The easiest way to prevent that is to have a local file that augments the existing definition:
```ts
interface CanvasRenderingContext2D {
	clip(path: Path2D, fillRule?: CanvasFillRule): void;
    fill(path: Path2D, fillRule?: CanvasFillRule): void;
    isPointInPath(path: Path2D, x: number, y: number, fillRule?: CanvasFillRule): boolean;
    isPointInStroke(path: Path2D, x: number, y: number): boolean;
}
```
[Playground link](http://www.typescriptlang.org/play/#src=interface%20CanvasRenderingContext2D%20%7B%0D%0A%09clip(path%3A%20Path2D%2C%20fillRule%3F%3A%20CanvasFillRule)%3A%20void%3B%0D%0A%20%20%20%20fill(path%3A%20Path2D%2C%20fillRule%3F%3A%20CanvasFillRule)%3A%20void%3B%0D%0A%20%20%20%20isPointInPath(path%3A%20Path2D%2C%20x%3A%20number%2C%20y%3A%20number%2C%20fillRule%3F%3A%20CanvasFillRule)%3A%20boolean%3B%0D%0A%20%20%20%20isPointInStroke(path%3A%20Path2D%2C%20x%3A%20number%2C%20y%3A%20number)%3A%20boolean%3B%0D%0A%7D%0D%0A%0D%0Avar%20x%20%3D%20(null%20as%20HTMLCanvasElement).getContext('2d')%3B%0D%0Ax.fill(null%20as%20Path2D)%3B)

Note that this file needs to *not* be a module (i.e. have top-level `export` or `import` declarations). I recommend just naming it `globalfixes.d.ts` or similar and just keeping workaround definitions in it.

**Comment by nhrones** [comment]
That is a method I've used in the past. Now, I can't seem to find any combination of file location/ tsconfig.json settings that work any longer. As it is, the files that I've modified in the vscode location <C:\Program Files (x86)\Microsoft VS Code\resources\app\extensions\node_modules\typescript\lib> generate many compile errors during the build task, but these errors 'clear' automatically as the cursor lands on the line with the error.  Very strange behavior indeed.  From what I've read, It may be related to the fact that my UI projects don't have any node_modules or package.json file. I do have several node file-server/socket-server projects that have no issues at all.
